File: workspace/exercise/w3/main.py
import heapq

chain_task = {}
completed_tasks = set()

def t1():
    print("t1 x++")
    activate_task(Task(t2, 3))
    print("t1 x++ (finish t1)")

def t2():
    print("t2 x--")
    activate_task(Task(t3, 2))
    print("t2 x-- (finish t2)")

def t3():
    print("t3 x++")
    print("t3 x-- (finish t3)")

chain_task = {
    t1: t2,
    t2: t3
}

class Task:
    def __init__(self, func, priority):
        self.func = func
        self.priority = priority

    def run(self):
        if self.func in chain_task:
            dep = chain_task[self.func]
            if dep not in completed_tasks:
                activate_task(Task(dep, self.priority + 1))
                activate_task(Task(self.func, self.priority))
                return
        self.func()
        completed_tasks.add(self.func)
        terminate_task()

q = []
s = []

def activate_task(t):
    if not s or t.priority > s[-1].priority:
        s.append(t)
        t.run()
    else:
        heapq.heappush(q, (-t.priority, t))

def terminate_task():
    if s:
        s.pop()
    if q and s:
        top_priority, top_task = q[0]
        if -top_priority > s[-1].priority:
            heapq.heappop(q)
            s.append(top_task)
            top_task.run()

def main():
    activate_task(Task(t1, 1))
    while q:
        _, next_task = heapq.heappop(q)
        next_task.run()

if __name__ == "__main__":
    main()

File: workspace/exercise/w3/main.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>

using namespace std;

static void t1();
static void t2();
static void t3();

struct Task {
    void (*func)();
    int priority;
    void run();
};

map<void(*)(), void(*)()> chain_task;
set<void(*)()> completed_tasks;

class MyStack {
    Task* arr;
    int cap;
    int topIndex;
public:
    MyStack(int c=1000) : cap(c), topIndex(-1) { arr = new Task[cap]; }
    ~MyStack() { delete[] arr; }
    bool empty() const { return topIndex < 0; }
    Task& top() { return arr[topIndex]; }
    void push(const Task& t) {
        if(topIndex+1<cap) {
            topIndex++;
            arr[topIndex] = t;
        }
    }
    void pop() { if(!empty()) topIndex--; }
};

class MyMaxHeap {
    vector<pair<int, Task>> data;
    void heapifyUp(int idx) {
        while(idx>0) {
            int p=(idx-1)/2;
            if(data[idx].first>data[p].first) {
                auto temp=data[idx]; data[idx]=data[p]; data[p]=temp;
                idx=p;
            } else break;
        }
    }
    void heapifyDown(int idx) {
        while(true) {
            int l=2*idx+1, r=2*idx+2, largest=idx;
            if(l<(int)data.size() && data[l].first>data[largest].first) largest=l;
            if(r<(int)data.size() && data[r].first>data[largest].first) largest=r;
            if(largest!=idx) {
                auto temp=data[idx]; data[idx]=data[largest]; data[largest]=temp;
                idx=largest;
            } else break;
        }
    }
public:
    bool empty() const { return data.empty(); }
    pair<int,Task> top() { return data[0]; }
    void push(const pair<int,Task>& val) {
        data.push_back(val);
        heapifyUp((int)data.size()-1);
    }
    void pop() {
        if(!data.empty()) {
            data[0] = data[data.size()-1];
            data.pop_back();
            if(!data.empty()) heapifyDown(0);
        }
    }
};

MyStack s;
MyMaxHeap q;

void activate_task(const Task& t);
void terminate_task();

void Task::run() {
    if(chain_task.find(func)!=chain_task.end()) {
        auto dep=chain_task[func];
        if(completed_tasks.find(dep)==completed_tasks.end()) {
            activate_task({dep,priority+1});
            activate_task({func,priority});
            return;
        }
    }
    func();
    completed_tasks.insert(func);
    terminate_task();
}

void t1() {
    cout<<"t1 x++"<<endl;
    activate_task({t2,3});
    cout<<"t1 x++ (finish t1)"<<endl;
}
void t2() {
    cout<<"t2 x--"<<endl;
    activate_task({t3,2});
    cout<<"t2 x-- (finish t2)"<<endl;
}
void t3() {
    cout<<"t3 x++"<<endl;
    cout<<"t3 x-- (finish t3)"<<endl;
}

void activate_task(const Task& t) {
    if(s.empty() || t.priority>s.top().priority) {
        s.push(t);
        s.top().run();
    } else {
        q.push({t.priority,t});
    }
}

void terminate_task() {
    if(!s.empty()) {
        s.pop();
    }
    if(!q.empty() && !s.empty()) {
        auto top_pair=q.top();
        if(top_pair.first>s.top().priority) {
            q.pop();
            s.push(top_pair.second);
            s.top().run();
        }
    }
}

int main(){
    // chain_task[t1]=t2;
    // chain_task[t2]=t3;
    activate_task({t1,1});
    while(!q.empty()){
        auto next_pair=q.top(); 
        q.pop();
        next_pair.second.run();
    }
    return 0;
}
